Extend multi selection by adding new mode "circle"

dopenguin opened the pull request:
## Summary

Add the possibility to select multiple features via a circle instead of a box.

This also deprecates the configuration parameter `boxSelect` in favour of the new parameter `multiSelect` which can be set to `'box'` or `'circle'`.

## Instructions for local reproduction and review

`npm run snowbox` and use CTRL / CMD and click once to start the circle and a second time to finish the selection.

## Relevant tickets, issues, et cetera

#124 has to be merged before this

warm-coolguy commented on packages/plugins/Gfi/README.md:
```diff
@@ -16,13 +16,14 @@ The GFI plugin can be used to fetch and optionally display GFI (GetFeatureInfo)
 | gfiContentComponent | VueConstructor? | Allows overriding the GfiContent.vue component for custom design and functionality. Coding knowledge is required to use this feature, as any implementation will have to rely upon the VueX store model. Please refer to the implementation. |
 | maxFeatures | number? | Limits the viewable GFIs per layer by this number. The first n elements are chosen arbitrarily. Useful if you e.g. just want one result, or to limit an endless stream of returns to e.g. 10. Infinite by default. |
 | mode | enum["bboxDot", "intersects"]? | Method of calculating which feature has been chosen by the user. `bboxDot` utilizes the `bbox`-url parameter using the clicked coordinate while `intersects` uses a `Filter` to calculate the intersected features. Layers can have their own `gfiMode` parameter which would override this global mode. To apply this, add the desired value to the parameter in the ...
+| multiSelect | enum["box", "circle"]? | If configured, multiple features can be selected at once by using the modifier key (CTRL on Windows or Command on macOS) and dragging the mouse. If set to `'box'`, the selection will be done in a box. If set to `'circle'`, the selection will be done in a circle beginning from the first click. In this case, the user has to click a second time to end the sel ...
```
While I understand that this feature is mostly used by clerks in an office/desktop environment, it lacks mobile support and is thus technically inaccessible on some devices. I wonder how we should treat that.

Note in documentation that it is meant for desktop use only and should not be used in citizen-facing clients? Maybe even make an issue of it, albeit that may be overzealous since afaik nobody needs this feature so far.

dopenguin replied:
Also, see https://github.com/Dataport/polar/pull/211#discussion_r1922193778

dopenguin replied:
Adjusted documentation fef2366975e05ef55267b37d865c3dc22a6a8bb7

warm-coolguy commented on packages/plugins/Gfi/src/store/actions/setupMultiSelection.ts:
```diff
@@ -0,0 +1,91 @@
+        coordinateOrExtent: e.feature.getGeometry().getExtent(),
+        modifierPressed: true,
+      }).finally(() =>
+        // Needed so that no pin is set when finishing drawing the circle
+        setTimeout(() => {
+          // @ts-expect-error | internal hack to detect it in @polar/plugin-pins
+          circleDraw._isMultiSelect = false
+        }, 250)
```
Does 0 suffice since all we need to do is get at the end of the execution chain or do we actually need the 250ms as timeout until an asynchronous request is done? In the latter case, I do not consider this a valid solution. We'd probably need a pins getter that returns an awaitable method that resolves on completing all current loading operations.

dopenguin replied:
IIRC I tested it with 0 or time lower than 250ms. I'll see what I can do to have a cleaner solution

dopenguin replied:
`Draw` offers a configuration parameter `stopClick` 0c38d86a95c04d600c35febe3ebcf96e2bb0971e

warm-coolguy commented on packages/plugins/Gfi/src/store/actions/setupMultiSelection.ts:
```diff
@@ -0,0 +1,91 @@
+const circleDraw = new Draw({
+  source: new VectorSource(),
+  type: 'Circle',
+  style: new Style({
+    stroke: new Stroke({ color: '#118bee', width: 2 }),
+    fill: new Fill({ color: [17, 139, 238, 0.6] }),
+  }),
```
I'd prefer the multiSelect circle to look more like the multiSelect dragBox.

dopenguin replied:
Sure! I'll change the styling

dopenguin replied:
Adjusted ea992a6ea5c7b226dc82aa6008cbfbb4dc2b668f

warm-coolguy commented on packages/plugins/Gfi/src/store/actions/setupMultiSelection.ts:
```diff
@@ -0,0 +1,91 @@
+      })
+    )
+    map.addInteraction(dragBox)
+  } else if (multiSelect === 'circle') {
```
I'd very much like the circle to behave as the box does in that the dragging operation is continued while the mouse is held down and ends on the `pointerup` or `pointercancel` event, since that is the "default OL behaviour" in the DragBox and thus probably expectable here, too. Also I fathom this to be the more mobile-like user interaction, should we ever come to the point where users may drag on mobile devices by e.g. clicking a button placed somewhere beforehand.

dopenguin replied:
I've fiddled around a bit and I don't seem to find a configuration of [Draw](https://openlayers.org/en/v9.2.4/apidoc/module-ol_interaction_Draw-Draw.html) that allows such interaction possibilities.

I've also realized that [DragBox](https://openlayers.org/en/v9.2.4/apidoc/module-ol_interaction_DragBox-DragBox.html) is not implemented for mobile usage. If we intend to support that, we'd have to either extend OL directly or implement something ourselves

warm-coolguy replied:
What do you think about trashing the DragBox then, and instead going for a box draw analoguous to the circle draw? https://openlayers.org/en/latest/examples/draw-shapes.html That, too, would even out the drawing experience.

dopenguin replied:
Sounds like a good idea UX-wise! Will implement that!

dopenguin replied:
Changed 17f91e01107ac37435b835e1782463940e5538a8

warm-coolguy commented on packages/plugins/Pins/src/store/index.ts:
```diff
@@ -53,8 +53,8 @@ export const makeStoreModule = () => {
             (interaction) =>
-              // these indicate other interactions are expected now
-              interaction instanceof Draw ||
+              // @ts-expect-error | internal hack to detect it from @polar/plugin-gfi
+              (interaction instanceof Draw && interaction._isMultiSelect) ||
```
This should be `or`ed instead of being `and`ed. Currently, while `Draw` is active, the marker is e.g. placed on the center of any circle drawn.

dopenguin replied:
The problem with simply adding an `or` here is that then the Pin can never be placed. I am not quite sure on how to handle this and will take a closer look at the end, maybe with a clearer idea

dopenguin replied:
Found a solution 9170d0fccd31bf1b3fc16c11cf50c20440decd38

warm-coolguy replied:
No crucially different idea comes to mind, and obviously we've had that problem before ...

Would you please write an issue/ticket regarding this? While I don't reckon there'll be a wildly different solution, we can at least standardize and document the flag for future plugins. I'm e.g. thinking about a `draw._isHeldByPolarPlugin = "Measure"`; this way, `Boolean(draw._isHeldByPolarPlugin)` can be used to determine whether a draw comes from a POLAR plugin, and `draw._isHeldByPolarPlugin === localPluginName` can be used to determine whether it's the current one ...

However, I consider this a separate effort. And before we double down on it, we should lock ourselves in the thinking chamber and think really hard about it. After you made a issue/ticket, 🎩 .

dopenguin replied:
I've added an internal ticket 👍

warm-coolguy reviewed (changes requested):
🏓 @dopenguin

dopenguin commented:
> 🏓 @dopenguin

🏓 @warm-coolguy

warm-coolguy commented:
> > 🏓 @dopenguin
> 
> 🏓 @warm-coolguy

🏓 @dopenguin 

Although #124 will soon be a blocker, there are some little leftovers to do. After #124 is merged, I will do another functionality check. I didn't verify the current branch state regarding that and will re-check at the end.

dopenguin commented:
> > > 🏓 @dopenguin
> > 
> > 
> > 🏓 @warm-coolguy
> 
> 🏓 @dopenguin
> 
> Although #124 will soon be a blocker, there are some little leftovers to do. After #124 is merged, I will do another functionality check. I didn't verify the current branch state regarding functionality and will re-check at the end.

We can the rest of the PR after #124 is done.

🏓 @warm-coolguy

dopenguin commented:
> > > > 🏓 @dopenguin
> > > 
> > > 
> > > 🏓 @warm-coolguy
> > 
> > 
> > 🏓 @dopenguin
> > Although #124 will soon be a blocker, there are some little leftovers to do. After #124 is merged, I will do another functionality check. I didn't verify the current branch state regarding functionality and will re-check at the end.
> 
> We can do the rest of the PR after #124 is done.
> 
> 🏓 @warm-coolguy

This PR is now ready receive its final review
:shipit: @warm-coolguy

warm-coolguy commented on packages/plugins/Gfi/README.md:
```diff
@@ -16,13 +16,14 @@ The GFI plugin can be used to fetch and optionally display GFI (GetFeatureInfo)
 | gfiContentComponent | VueConstructor? | Allows overriding the GfiContent.vue component for custom design and functionality. Coding knowledge is required to use this feature, as any implementation will have to rely upon the VueX store model. Please refer to the implementation. |
 | maxFeatures | number? | Limits the viewable GFIs per layer by this number. The first n elements are chosen arbitrarily. Useful if you e.g. just want one result, or to limit an endless stream of returns to e.g. 10. Infinite by default. |
 | mode | enum["bboxDot", "intersects"]? | Method of calculating which feature has been chosen by the user. `bboxDot` utilizes the `bbox`-url parameter using the clicked coordinate while `intersects` uses a `Filter` to calculate the intersected features. Layers can have their own `gfiMode` parameter which would override this global mode. To apply this, add the desired value to the parameter in the ...
+| multiSelect | enum["box", "circle"]? | If configured, multiple features can be selected at once by using the modifier key (CTRL on Windows or Command on macOS) and dragging the mouse. Can only be used in Desktop environments. If set to `'box'`, the selection will be done in a box. If set to `'circle'`, the selection will be done in a circle beginning from the first click. In this case, the user ...
```
It's actually "Shift" on Windows. This makes the toggling de-/selection behaviour confusing since such behaviour is associated with CTRL. My first reaction was to report the behaviour as a bug. Users will probably do the same.

Proposal:

* Selection with "Shift" is purely additive, users have to click somewhere else for deselection
* Selection with "Shift+Strg" has the behaviour currently seen in "Shift" (toggling de-/selection)

This mimics how file selection works on the Windows Explorer, which most users should be familiar with, if they are familiar with any selection scheme. Is it wildly different on MacOS?

warm-coolguy replied:
>If set to `'circle'`, the selection will be done in a circle beginning from the first click. In this case, the user has to click a second time to end the selection.

This is not true on Windows, where the `circle` behaves in the same fashion as the `box` does, id est "Hold Shift, mouse down, drag circle to desired size, mouse up" encompasses the whole operation. On the other hand, acting as described ("Hold Shift, Click, Drag, Click") does not work.

Is it different on MacOS?

warm-coolguy commented on packages/plugins/Gfi/README.md:
```diff
@@ -16,13 +16,14 @@ The GFI plugin can be used to fetch and optionally display GFI (GetFeatureInfo)
 | activeLayerPath | string? | Optional store path to array of active mask layer ids. If used with `LayerChooser`, setting this to `'plugin/layerChooser/activeMaskIds'` will result in an info text in the GFI box, should no layer be active. If used without `LayerChooser`, the active mask layers have to be provided by another plugin or the client. If not set, the GFI plugin may e.g. show an empty li ...
 | afterLoadFunction | function (featuresByLayerId: Record<string, GeoJsonFeature[]>): Record<layerId, GeoJsonFeature[]>? | This method can be used to extend, filter, or otherwise modify a GFI result. |
-| boxSelect | boolean? | If set to `true`, multiple features can be selected at once by using the modifier key (CTRL on Windows or Command on macOS) and dragging the mouse. Similar to `gfi.directSelect`, features can be added and removed by selection / unselecting them. The features need to be distinguishable by their properties for the functionality to properly work. Does not work together with  ...
+| boxSelect | boolean? | **This parameter has been deprecated. Please use `multiSelect` to `'box'` instead.** If set to `true`, multiple features can be selected at once by using the modifier key (CTRL on Windows or Command on macOS) and dragging the mouse. Can only be used in Desktop environments. Similar to `gfi.directSelect`, features can be added and removed by selection / unselecting them. T ...
```
As on the new parameter (see other comment), the documentation is currently wrong regarding CTRL.

warm-coolguy reviewed (changes requested):
>This PR is now ready receive its final review
>:shipit: @warm-coolguy

You can't shipit me since I am not the one to merge this pull request. Shipitting me would have to happen on a pull request I am the assignee of.

🏌️ @dopenguin

warm-coolguy commented:
@dopenguin The type `MeasureConfiguration` sneaked into our core types with the last PR merge. Since it's topical, please remove it here.

dopenguin commented:
> > This PR is now ready receive its final review
> > :shipit: @warm-coolguy
> 
> You can't shipit me since I am not the one to merge this pull request. Shipitting me would have to happen on a pull request I am the assignee of.
> 
> 🏌️ @dopenguin

Removed the type 8b29f6acab447492a7d7ba811e95709d3fc73414 and changed the behaviour as described 42faf6390900820d93f2297d2347f234edeb6207.

It now works on Mac, can you test the same for Windows and if it is the way we talked?

🏓 @warm-coolguy

warm-coolguy commented on packages/plugins/Gfi/src/store/actions/setupMultiSelection.ts:
```diff
@@ -0,0 +1,94 @@
+        '@polar/plugin-gfi: Configuration parameter "boxSelect" has been deprecated. Please use the new parameter "multiSelect" set to "box" instead.'
+      )
+    }
+    if (multiSelect !== 'circle') {
+      drawOptions.geometryFunction = createBox()
+    }
```
```suggestion
    if (multiSelect !== 'circle') {
      drawOptions.geometryFunction = createBox()
    }
    else {
      delete drawOptions.geometryFunction;
    }
```
Just in case we ever have switchable select logic at run-time.

dopenguin replied:
Done dbb2cec92897d2881a272bebaaf6f3b3e506fce6

warm-coolguy reviewed (approved):
> > > This PR is now ready receive its final review
> > > :shipit: @warm-coolguy
> > 
> > 
> > You can't shipit me since I am not the one to merge this pull request. Shipitting me would have to happen on a pull request I am the assignee of.
> > 🏌️ @dopenguin
> 
> Removed the type [8b29f6a](https://github.com/Dataport/polar/commit/8b29f6acab447492a7d7ba811e95709d3fc73414) and changed the behaviour as described [42faf63](https://github.com/Dataport/polar/commit/42faf6390900820d93f2297d2347f234edeb6207).
> 
> It now works on Mac, can you test the same for Windows and if it is the way we talked?
> 
> 🏓 @warm-coolguy

:shipit: @dopenguin It works. I have one last remark in https://github.com/Dataport/polar/pull/211#pullrequestreview-2584318633 and the PR is good to go.
